Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -62,78 +164,188 @@ public void Initialize()
+
             this.logger.LogTrace("(-)");
         }
 
-        public IEnumerable<BlockResult> Execute(CancellationToken cancellationToken)
+        /// <summary>
+        /// Dispose components in <see cref="ConsensusLoop"/>.
+        /// </summary>
+        public void Stop()
         {
-            while (true)
+            this.asyncLoop?.Dispose();
+        }
+
+        /// <summary>
+        /// A puller method that will continuously loop and ask for the next block  in the chain from peers.
+        /// The block will then be passed to the consensus validation. 
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token that will stop the loop.</param>
+        /// <returns>The task the loop will be running under.</returns>
+        private Task PullerLoop(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
             {
-                yield return this.ExecuteNextBlock(cancellationToken);
+                try
+                {
+                    BlockItem item = new BlockItem();
+
+                    using (new StopwatchDisposable(o => this.Validator.PerformanceCounter.AddBlockFetchingTime(o)))
+                    {
+                        item.Block = this.Puller.NextBlock(cancellationToken);
+                    }
+
+                    this.logger.LogTrace("Block received from puller.");
+                    this.AcceptBlock(item);
+
+                }
+                catch (Exception ex)
+                {
+                    if (ex is OperationCanceledException)
+                    {
+                        if (this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+                            return Task.FromException(ex);
+                    }
+
+                    // TODO Need to revisit unhandled exceptions in a way that any process can signal an exception has been
+                    // thrown so that the node and all the disposables can stop gracefully.
+                    this.logger.LogCritical(new EventId(0), ex, "Consensus loop at Tip:{0} unhandled exception {1}", this.Tip, ex.ToString());
+                    NLog.LogManager.Flush();
+                    throw;
+                }
             }
+
+            return Task.CompletedTask;
         }
 
-        public BlockResult ExecuteNextBlock(CancellationToken cancellationToken)
+        /// <summary>
+        /// A method that will accept a new block to the node.
+        /// The block will be validated and the <see cref="CoinView"/> db will be updated.
+        /// If its a new block that extends the chain then the <see cref="ConcurrentChain.Tip"/> will be set.
+        /// </summary>
+        /// <param name="item">The block to validate</param>
+        public void AcceptBlock(BlockItem item)
         {
             this.logger.LogTrace("()");
 
-            BlockResult result = new BlockResult();
-            try
+            lock (this.consensusLock)
```

Comment:
I have three problems here. First, the critical section is enormous. It can take very long time. Second, you can't await inside lock(). Third, as the section is very long, another thread can be stuck here without cancellation. 

All these problems could be solved by not using lock(), but using something else. And since it seems we will want to await, the newly merged AsyncLock with cancellation token is probably the way to go.

Reply:
Yes, I am still not sure how we will make this async I have two ideas (either AsyncLock or a dedicated loop) but in any case only 2 (or 3 threads) will be ever accessing this code (puller and both the miners) so not a big deal (right now).

For that reason the fact the critical section is big is not such a big deal.

Reply:
that is fine, the problem is that lock() is not very good thing when your critical section is so large
that asynclock is based on semaphore and that's very suitable here
i'm talking about the (in)efficiency of entering the lock 

so yes, i agree it is not a big deal, but the fix is absolutely trivial and it would look much better
instead of 
`lock (x)` we use `using (x.LockAsync(cancellationtokenhere))` for some `private AsyncLock x = new AsyncLock();`